_posts/2017-07-29-ruby-oop.md

```diff
@@ -0,0 +1,383 @@
+
+- `attr_reader :var` - Creates a read-only method for attribute `@var`.
+- `attr_writer :var` - Creates a write-only method for attribute `@var`.
+- `attr_accessor :var` - Creates both kinds of methods to read and write to this attribute.
+
+So, our `Point` can be described simply as:
+```ruby
+# A location in 2D space.
+class Point
+  attr_reader :x
+
+  attr_reader :y
+
+  def initialize(x, y)
+    @x = x
+    @y = y
+  end
+end
+```
+> In our case, we will only need to ever read these attributes, so I just used `attr_reader`. `attr_accessor` would have also worked here.
+
+Playing with our new readers:
+```ruby
+point = Point.new(1.5, -3.0)
+puts point.x #=> 1.5
+puts point.y #=> -3.0
+```
+> You can read more about attributes, readers, setters, and more [here](http://ruby-for-beginners.rubymonstas.org/writing_classes/instance_variables.html).
+
+What else would be helpful to do with our point, or perhaps any other point? Well, with any two points, you can calculate the distance between them.
+
+Ruby has a rich standard library, which includes some common mathematical functions. We need the pythagorean theorem here to calculate the distance between two points, and if we do some searching we can see Ruby already provides a method for us to calculate this instead of having to implement it ourselves:
+
+[https://ruby-doc.org/core-2.4.0/Math.html#method-c-hypot](https://ruby-doc.org/core-2.4.0/Math.html#method-c-hypot)
+```ruby
+Math.hypot(3, 4) #=> 5.0
+```
+
+Let's implement it to our `Point`. (Note: I've omitted the rest of the class we already wrote to save space, ad nauseum)
+```ruby
+# A location in 2D space.
+class Point
+  # Calculates the distance to another Point object
+  def distance(other)
+    x_difference = x - other.x
+    y_difference = y - other.y
+
+    Math.hypot x_difference, y_difference
+  end
+end
+```
+> Here, the `x` in `x_difference = x - other.x` is actually calling this objects *own* `attr_reader` method internally that fetches its `@x` attribute. In another language, you may write something like `self.x` to refer to this value. You can do that here too in Ruby, but in most cases you can omit `self.`, because Ruby understands what you mean.
+
+Our new `#distance` method will take another `Point` object as an argument, read its `@x` and `@y` attributes, and do the maths to return our result implicitly by being the last line of the method.
+
+Let's try it out:
+```ruby
+point_a = Point.new(1.5, -3.0)
+point_b = Point.new(2, 1.0)
+
+puts point_a.distance point_b #=> 4.031128874149275
+```
+
+**We can now:**
+
+1. Describe a `Point`.
+2. Measure the distance between any two `Point`s.
+
+### Line
+
+Let's refer back to our definition of a `Line`.
+
+```ruby
+# A finite segment (or distance) between two points.
+class Line
+end
+```
+
+From what we know, we can now give our `Line`s two attributes. We'll call them `start` and `stop` (I would use the word `end` here, but this is a Ruby keyword and it could get confusing.)
```

Comment:
Maybe add to say that you cannot use Ruby keywords as variable names, not that it's confusing to use. (Yes, there are symbol workarounds, but I think it's better to talk about this rather than leave it vague)